### PR TITLE
Fix wildcard reads + chunking

### PR DESCRIPTION
**Pull request by mrjerryjohns**
Using the Matter REPL as a validation tool, this PR fixes a number of critical bugs in both the server-side logic as well as the Python APIs to validate all permutations of wildcard reads as well as basic chunking.
    
On the server side, it fixes:
    
- The `AttributeExpandPathIterator` was incorrectly using `emberAfClusterIndex` to iterate over clusters on an endpoint, but that isn't its API contract. That function returns something else completely. Fixed that implementation as well as created a new, better named function called `emberAfClusterIndexInMatchingEndpoints` that actually better describes what it does.
    
- In `ReadSingleClusterData`, we were chaining a number of MessageDef builder calls one after another and checking the error at the end (by calling `GetError()`). However, this resulted in the original error encountered in the first call in that chain being lost and replaced with something else by the time it got to the end. This was especially problematic during chunking since the errors `CHIP_ERROR_BUFFER_TOO_SMALL` and `CHIP_ERROR_NO_MEMORY` are what the higher level calls in `Engine::BuildSingleReportDataAttributeReportIBs` expect to correctly handle chunking. Since the error got lost and converted, it resulted in the Engine treating it like a critical failure and stopping the generation of reports entirely.

- Many places where we were assuming we could close out DataIB after encountering a TLV full error when writing data, which was incorrectly and resulted in the subscription failing. Instead, doing the correct reservation of space for these end-of-container tags was done in the fix to prevent encountering any subsequent errors on closure.
    
On the client side in Python:
    
- Reading `Clusters.OnOff.Attributes.SampleMfgSpecificAttribute0x00000x1049` broke because in the Python code, we were generating attribute IDs that were 16-bit values instead of 32-bits. This was a combination of not generating them correctly, as well as the C++ bindings expecting a c_uint16 instead of c_uint32
    
- Fixed up `ChipDeviceController.ReadAttribute` to actually work correctly with wildcards instead of just erroring out due to parsing issues.
    
- Made the data returned by ReadAttribute actually return the value as strongly typed values, making it easier to read.

Validation:
    
- Tested all wildcard read variations using the Python ReadAttribute API in the REPL and ensured it completed successfully, including reading **every single attribute on the device!** and validating that each and every attribute was of the right type.

- Created a definitive, robust chunking validation unit test that artificially limits the size of a packet buffer and sweeps through a range of packet buffer sizes to empirically fit in N to N+1 attributes in a report, exposing all corner cases. This was used to not only ensure the above bugs were caught by the test, but to also validate the fixes.
    
- Added all variations as tests to the mobile device test suite.

### CODE REVIEW

File: src/app/util/attribute-storage.h [hunk 1]
```diff
@@ -131,11 +131,13 @@ bool emAfMatchCluster(EmberAfCluster * cluster, EmberAfAttributeSearchRecord * a
 bool emAfMatchAttribute(EmberAfCluster * cluster, EmberAfAttributeMetadata * am, EmberAfAttributeSearchRecord * attRecord);
 
 EmberAfCluster * emberAfFindClusterInTypeWithMfgCode(EmberAfEndpointType * endpointType, chip::ClusterId clusterId,
-                                                     EmberAfClusterMask mask, uint16_t manufacturerCode);
+                                                     EmberAfClusterMask mask, uint16_t manufacturerCode, uint8_t * index = nullptr);
 
 EmberAfCluster * emberAfFindClusterInType(EmberAfEndpointType * endpointType, chip::ClusterId clusterId, EmberAfClusterMask mask);
 
-// This function returns the index of cluster for the particular endpoint.
+// For a given cluster and mask, retrieves the list of endpoints that contain the matching cluster and returns the index
```
Comment (mrjerryjohns) [inline src/app/util/attribute-storage.h:138]:
```suggestion
// For a given cluster and mask, retrieves the list of endpoints sorted by endpoint that contain the matching cluster and returns the index
```

File: src/controller/python/test/test_scripts/cluster_objects.py [hunk 2]
```diff
@@ -96,54 +96,52 @@ async def SendWriteRequest(cls, devCtrl):
...
-            (0, Clusters.Basic.Attributes.ProductName),
             (0, Clusters.Basic.Attributes.ProductID),
-            (0, Clusters.Basic.Attributes.UserLabel),
-            (0, Clusters.Basic.Attributes.Location),
             (0, Clusters.Basic.Attributes.HardwareVersion),
-            (0, Clusters.Basic.Attributes.HardwareVersionString),
-            (0, Clusters.Basic.Attributes.SoftwareVersion),
-            (0, Clusters.Basic.Attributes.SoftwareVersionString),
         ]
+        await devCtrl.ReadAttribute(nodeid=NODE_ID, attributes=req)
```
Comment (erjiaqing) [inline src/controller/python/test/test_scripts/cluster_objects.py:114]: validate that we got non-empty responses here?

File: src/controller/python/chip/ChipDeviceCtrl.py [hunk 3]
```diff
@@ -422,15 +422,19 @@ async def ReadAttribute(self, nodeid: int, attributes: typing.List[typing.Union[
...
-            ():                                         Endpoint = *,           Cluster = *,          Attribute = *
-            (Clusters.ClusterA):                        Endpoint = *,           Cluster = specific,   Attribute = *
-            (Clusters.ClusterA.AttributeA):             Endpoint = *,           Cluster = specific,   Attribute = specific
-            (endpoint, Clusters.ClusterA):              Endpoint = specific,    Cluster = specific,   Attribute = *
             (endpoint, Clusters.ClusterA.AttributeA):   Endpoint = specific,    Cluster = specific,   Attribute = specific
+            (endpoint, Clusters.ClusterA):              Endpoint = specific,    Cluster = specific,   Attribute = *
+            (Clusters.ClusterA.AttributeA):             Endpoint = *,           Cluster = specific,   Attribute = specific
+            endpoint:                                   Endpoint = specific,    Cluster = *,          Attribute = *
+            Clusters.ClusterA:                          Endpoint = *,           Cluster = specific,   Attribute = *
+            None:                                       Endpoint = *,           Cluster = *,          Attribute = *
```
Comment (erjiaqing) [inline src/controller/python/chip/ChipDeviceCtrl.py:430]: I would use `('*')` for a convenient represent for a path with all wildcard components here with better readability than `None`, can you add a note here?

**Comment by mrjerryjohns** [comment]
Unfortunately, there are a bunch more bugs on the server side that the tests I've added to Python are un-earthing. So, will need to fix these before this PR is in a shape to merge.

**Review by andy31415 — APPROVED** [review]

**Comment by andy31415** [comment]
fast track: change up for a while, created by a domain owner. Most changes are python-related which is not affecting main functionality.
